numpy/ma/core.py

```diff
@@ -4984,23 +4877,51 @@ def mean(self, axis=None, dtype=None, out=None):
 
         """
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = super(MaskedArray, self).mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    if any(zerocounts):
+                        out.__setmask__(zerocounts)
+                    else:
+                        out.__setmask__(nomask)
+                return out
+            # mask in case of empty slices
+            if any(zerocounts):
+                ret = ret.view(MaskedArray)
+                ret.__setmask__(zerocounts)
+            return ret
+
+        # code largely copied from _methods.mean
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
I first thought so too, but I saw that a unit test makes a copy ([here](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_umath.py#L1172), and the comment suggests it's intentional), and I also see that at least astropy's `Quantity` also sometimes returns a copy, [here](https://github.com/astropy/astropy/blob/c494d392da899dd0f01855c9a951a5fd29319635/astropy/units/quantity.py#L436). 

I can't say I totally understand the reason, but it does seem like setting `out=outarr, subok=False` is a way of writing to out without without calling `__array_prepare__`, and that this call also preserves out's subtype.